### PR TITLE
Minimal_generating_set/small_generating_set -- local cases

### PR DESCRIPTION
**Pull request by afkafkafk13**
PR contains:
- minimal_generating_set for localizations at k-points
- small_generating_set for localizations at k-points and localizations at powers of elements

Standard Basis for shifted_ideal w.r.t. local ordering is stored in dictionary as in the global case.

Obsoletes PR #2327
Resolves Issue #2295

### CODE REVIEW

**Comment by afkafkafk13** [comment]
@fingolfin : What am I doing wrong causing some tests to not have started?

**Comment by benlorenz** [comment]
Close and reopen will restart CI (with a fresh merge-commit with master).

The required status checks have been updated for the changes from #2533.

File: src/Rings/mpoly-localizations.jl
Comment (simonbrandhorst) [inline src/Rings/mpoly-localizations.jl:3054]: Minimal with respect to inclusion or cardinality?

File: src/Rings/mpoly-localizations.jl
Comment (simonbrandhorst) [inline src/Rings/mpoly-localizations.jl:3093]: Do we have an interface for this? If we have, would you please use it?
Or we could introduce one? (in a different pr) like `has_groebner_basis`?

File: src/Rings/mpoly-localizations.jl
Comment (simonbrandhorst) [inline src/Rings/mpoly-localizations.jl:3096]: Similar here.

File: src/Rings/mpoly-localizations.jl
Comment (simonbrandhorst) [inline src/Rings/mpoly-localizations.jl:3098]: and here

File: src/Rings/mpoly-localizations.jl
Comment (simonbrandhorst) [inline src/Rings/mpoly-localizations.jl:3100]: and here

**Review by simonbrandhorst — APPROVED** [review]
The interface looks good to me. Can't say much about the math involved. But I assume you know what you do.

File: src/Rings/mpoly-localizations.jl
Comment (afkafkafk13) [reply src/Rings/mpoly-localizations.jl:3054]: This is a fixed technical term. It is minimal as in minimal presentation of R/I (exactly like in the graded case), which in turn implies minimal w.r.t. cardinality.

File: src/Rings/mpoly-localizations.jl
Comment (afkafkafk13) [reply src/Rings/mpoly-localizations.jl:3093]: No, we do not have an interface for this in the local case. And the function in question is located at a very low level in the hierarchy of code so that I see it as no problem here.

At some point in the future, we need something to properly interact with the standard basis of the shifted ideal in the localized case. But this reaches into the localized ring code and is not intended as content of this PR (parts of which could however serve as blueprint for this).

Comment (afkafkafk13) [reply src/Rings/mpoly-localizations.jl:3093]: P.S: This is a comment also to the whole sequence of commands to which you commented below.

File: src/Rings/mpoly-localizations.jl
Comment (HechtiDerLachs) [reply src/Rings/mpoly-localizations.jl:3054]: I would probably add a `note` saying that minimality is strict only in the case of localization at a rational point.

File: src/Rings/mpoly-localizations.jl
Comment (HechtiDerLachs) [inline src/Rings/mpoly-localizations.jl:3087]: This is already asserted by the signature of the method.

File: src/Rings/mpoly-localizations.jl
Comment (HechtiDerLachs) [reply src/Rings/mpoly-localizations.jl:3096]: I agree that such an interface would be nice to have. But there is none. This code is just poking in the internals and might not even be safe. But for the moment, I'm afraid, we can't do better. It's a bigger task that we have to address here in the beginning of August.

File: src/Rings/mpolyquo-localizations.jl
Comment (HechtiDerLachs) [inline src/Rings/mpolyquo-localizations.jl]: Shouldn't there also be a shift here? This method does not seem to do anything useful for now. Is this code still due to something I did?

File: test/Rings/mpolyquo-localizations.jl
Comment (HechtiDerLachs) [inline test/Rings/mpolyquo-localizations.jl]: I'm afraid, these tests are still too trivial. Since there is no shifting visible in the tested methods, I expect these tests to fail for ideals with more generators.

**Review by HechtiDerLachs — COMMENTED** [review]
Maybe I'm missing some things, but this doesn't look fully functional to me for now. 

For the `MPolyQuoLocalizedIdeal`s I propose to use the `oscar.pre_image_ideal` (sorry about the name of this method) to link this to the `MPolyLocalizedIdeal`s which have already been covered.

File: src/Rings/mpoly-localizations.jl
Comment (afkafkafk13) [reply src/Rings/mpoly-localizations.jl:3087]: So we can be sure that no other fields than AbstractAlgebra.Fields are floating around somewhere? I was not sure about that.

File: src/Rings/mpolyquo-localizations.jl
Comment (afkafkafk13) [reply src/Rings/mpolyquo-localizations.jl]: This is pulling the ideal in the localized quotient ring back to the localized ring, where we fall back to the method there. The shift is going on in the ring L not in Q -- to keep the ugly code at one specific place and not pour it over many files.

File: src/Rings/mpoly-localizations.jl
Comment (afkafkafk13) [reply src/Rings/mpoly-localizations.jl:3054]: It is also strict for localizations at prime ideal (seen als localized ring over a field extension) -- but we do not have functionality for this kind of ring in the backend. So I put a note there stating that it is only available for rational points.

**Comment by afkafkafk13** [comment]
> Maybe I'm missing some things, but this doesn't look fully functional to me for now.
> 
> For the `MPolyQuoLocalizedIdeal`s I propose to use the `oscar.pre_image_ideal` (sorry about the name of this method) to link this to the `MPolyLocalizedIdeal`s which have already been covered.

Thanks for the hint. I will change it accordingly.

Why is this command neither documented nor exported? I tried to use as few non-exported methods as possible.

File: test/Rings/mpolyquo-localizations.jl
Comment (afkafkafk13) [reply test/Rings/mpolyquo-localizations.jl]: You are right. More is in the next PR.

File: src/Rings/mpoly-localizations.jl
Comment (afkafkafk13) [reply src/Rings/mpoly-localizations.jl:3096]: I put a guard in to make sure that a non-local ordering will not be used -- this could however throw an error. 
But I did not want to go through the whole list I_shift.gb to find a local one. I_shift is cached internal data for an ideal in a localized ring after all.

File: src/Rings/mpolyquo-localizations.jl
Comment (afkafkafk13) [reply src/Rings/mpolyquo-localizations.jl]: I replaced it by pre_image_ideal. Thanks for the pointer.

**Comment by afkafkafk13** [comment]
> Maybe I'm missing some things, but this doesn't look fully functional to me for now.

I tested it (beyond the tests in test/Rings/ directory) and it is fully functional for the localizations at a rational point and parallels the graded code. However, I use knowledge on the internal workings of mstd and only provide an interface for that.

**Comment by HechtiDerLachs** [comment]
> Why is this command neither documented nor exported? I tried to use as few non-exported methods as possible.

It is supposed to be an internal method. I don't think the average user is interested in this. But it should be considered to be legitimate to use it internally for such purposes as here. At least that's what I wrote it for. 

But: For quotient rings there is some magic happening with `singular_assure` so that you can get a quotient ring on the singular side directly. This affects the outcome of `minimal_generating_set`: A minimal generating set of the `pre_image_ideal` might not be minimal modulo the `modulus` of the ring. So I'm afraid we have to be more careful here and do go via shiftings etc. again. The method for `MPolyQuoRing`s should provide us with samples how to bind to the singular backend correctly, then.

**Comment by afkafkafk13** [comment]
> 

I see your point and will take care about it.

File: src/Rings/mpolyquo-localizations.jl
Comment (HechtiDerLachs) [inline src/Rings/mpolyquo-localizations.jl]:
```suggestion
  J = pre_image_ideal(I)
```

File: src/Rings/mpolyquo-localizations.jl
Comment (HechtiDerLachs) [inline src/Rings/mpolyquo-localizations.jl]:
```suggestion
  J = pre_image_ideal(I)
```

File: src/Rings/mpolyquo-localizations.jl
Comment (HechtiDerLachs) [inline src/Rings/mpolyquo-localizations.jl:2235]:
```suggestion
@attr function small_generating_set(
```
Maybe we should turn all these into attributes?

**Review by HechtiDerLachs — APPROVED** [review]
Sorry, I think, I misread something earlier. This should all be rather fine.

**Comment by HechtiDerLachs** [comment]
Sorry, but I have to revoke my approval. It seemed to be working, but looking at the code, I couldn't see why, unless this was a coincidence. And indeed I found the following counterexample:
```
R, (x, y) = QQ[:x, :y]
I = ideal(R, x^6-y)
U = complement_of_point_ideal(R, [1, 1])
L = MPolyQuoLocRing(R, I, U)
J = ideal(L, [x-1, y-1])^2
oscar.minimal_generating_set(J)
```
The modulus suggests that the ring is effectively a polynomial ring in one variable `x`. Then the minimal generating set of the ideal `J` should consist of one generator, but my output is
```
julia> oscar.minimal_generating_set(J)
2-element Vector{MPolyQuoLocRingElem{QQField, QQFieldElem, QQMPolyRing, QQMPolyRingElem, MPolyComplementOfKPointIdeal{QQField, QQFieldElem, QQMPolyRing, QQMPolyRingElem}}}:
 y^2 - 2*y + 1
 6*x - y - 5
```

**Comment by afkafkafk13** [comment]
As I already wrote before your approval and the taking back thereof: 
I saw your point and I am working on a solution. 

Unfortunately, it turned out that the difficulties lie in functionality which is available for MPolyLocRing, but not for MPolyLocQuoRing (for good reasons). I know how to amend for that, but  it will take at least until monday.

File: src/Rings/mpolyquo-localizations.jl
Comment (lgoettgens) [reply src/Rings/mpolyquo-localizations.jl:2235]: Please add the type of the cached object after `@attr` to ensure type safety.

**Comment by afkafkafk13** [comment]
Thanks @HechtiDerLachs for pointing out the problem. I now do the locquo-case 'by hand' as there is not enough underlying functionality available. Please have look.

File: src/Rings/mpolyquo-localizations.jl
Comment (HechtiDerLachs) [inline src/Rings/mpolyquo-localizations.jl:2170]: Out of curiosity? Did the type assertions here cause any trouble? Otherwise: Why not keep them?

File: src/Rings/mpolyquo-localizations.jl
Comment (HechtiDerLachs) [inline src/Rings/mpolyquo-localizations.jl]:
```suggestion
@attr Vector{<:MPolyQuoLocRingElem} function minimal_generating_set(
```

File: src/Rings/mpolyquo-localizations.jl
Comment (HechtiDerLachs) [inline src/Rings/mpolyquo-localizations.jl]: This docstring is misplaced. Also, the function below should probably just be a synonym for `minimal_generating_set`.

**Review by HechtiDerLachs — APPROVED** [review]
This looks good. It's hard to really understand just from reading what the singular code is doing, but if the tests work out, it should be fine.

File: src/Rings/mpolyquo-localizations.jl
Comment (afkafkafk13) [reply src/Rings/mpolyquo-localizations.jl]: done.

File: src/Rings/mpolyquo-localizations.jl
Comment (afkafkafk13) [reply src/Rings/mpolyquo-localizations.jl:2235]: done

File: src/Rings/mpolyquo-localizations.jl
Comment (afkafkafk13) [reply src/Rings/mpolyquo-localizations.jl:2170]: Yes and no, it did cause trouble during the my many tests and debugging runs because of my incompetence to always feed suitable objects to it. The two imposed conditions suffice for us, so I left out the others.

File: src/Rings/mpolyquo-localizations.jl
Comment (afkafkafk13) [reply src/Rings/mpolyquo-localizations.jl]: done

File: src/Rings/mpolyquo-localizations.jl
Comment (afkafkafk13) [reply src/Rings/mpolyquo-localizations.jl]: No, the doctring has a Copy&Paste Error. The next commit amends this.
And no, this should not be an alias, because the other variant (which does it on foot)  is more clumsy and this one often suffices, if you do not require minimality, but only something reasonably small.

**Pull request merged**
